pkg/dsutil/set.go

```diff
@@ -0,0 +1,115 @@
+package dsutil
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"golang.org/x/exp/constraints"
+)
+
+// Set implements a set data structure based on built-in maps. This is not
+// optimized for large data, but to be convient.
+type Set[T constraints.Ordered] struct {
+	data map[T]struct{}
+}
+
+// Add puts a single value to the set. The set will be the same, if it already
+// contains the value.
+func (s *Set[T]) Add(value T) {
+	if s.data == nil {
+		s.data = map[T]struct{}{}
+	}
+
+	s.data[value] = struct{}{}
+}
+
+// Contains returns true, if the given value is part of the set.
+func (s *Set[T]) Contains(value T) bool {
+	if s.data == nil {
+		return false
+	}
+
+	_, found := s.data[value]
+	return found
+}
+
+// Remove removes the given value from the set. The set will be the same, if
+// the value is not part of it.
+func (s *Set[T]) Remove(value T) {
+	if s.data == nil {
+		return
+	}
+
+	delete(s.data, value)
+}
+
+// Substract removes every element from the given set from the set.
+func (s *Set[T]) Subtract(other *Set[T]) {
+	if s.data == nil {
+		return
+	}
+
+	for value := range other.data {
+		delete(s.data, value)
+	}
+}
+
+// Intersect removes all values from the set that are not in both sets.
+func (s *Set[T]) Intersect(other *Set[T]) {
+	for value := range s.data {
+		if !other.Contains(value) {
+			delete(s.data, value)
+		}
+	}
+}
+
+// Len returns the number of all values in the set.
+func (s Set[T]) Len() int {
+	return len(s.data)
+}
+
+// ToList converts the set into a slice. Since the set uses a map as an
+// underlying data structure, this will copy each value. So it might be memory
+// intensive. Also it sorts the slice to ensure a cosistent result.
+func (s *Set[T]) ToList() []T {
+	list := make([]T, 0, len(s.data))
+
+	if len(s.data) > 0 {
+		for v := range s.data {
+			list = append(list, v)
+		}
+	}
+
+	sort.Slice(list, func(i, j int) bool { return list[i] < list[j] })
+
+	return list
+}
+
+// Union adds each value from the given set to the set.
+func (s *Set[T]) Union(other Set[T]) {
+	for _, o := range other.ToList() {
```

Comment:
No way to add directly without creating the list firs?

Reply:
Good point. Fixed in aebc957f13af60fe7ad86a2864b10a43ad6a1726